### PR TITLE
Store anode wire hits in ALCT (ACLUT-3)

### PR DESCRIPTION
**Pull request by dildick**
#### PR description:

This PR stores the anode wire hits in the ALCT data format, so one can do detailed studies of the performance of the ALCT algorithm for Run-3 or Phase-2 (e.g. displaced muons, close-by muons, hadronic showers,...). It also allows for new logic to be developed aimed at enhancing the ALCT position resolution for Phase-2.

The PR is similar to #29233, and builds on top of https://github.com/cms-sw/cmssw/pull/30104 (which has yet to be merged). 

#### PR validation:

Tested with WF 20434.0 and on 100 single muon gun events

```
cmsDriver.py Configuration/Generator/python/SingleMuPt10_pythia8_cfi.py \
--fileout file:step1.root --mc \
--eventcontent FEVTDEBUG --datatier GEN-SIM --conditions auto:phase1_2021_realistic \
--beamspot Run3RoundOptics25ns13TeVLowSigmaZ --step GEN,SIM --geometry DB:Extended \
--era Run3 --python_filename SingleMu_GEN_SIM.py --no_exec   -n 10

cmsDriver.py step2.py --filein file:step1.root --fileout file:step2.root --mc \
--eventcontent FEVTDEBUG --datatier GEN-SIM-DIGI-L1 --conditions auto:phase1_2021_realistic \
--step DIGI:pdigi_valid,L1 --geometry DB:Extended \
--era Run3 --python_filename step2_DIGI_L1.py --no_exec   -n 10
```

Somewhat verbose output of the simulation of a single pattern that ultimately produced an ALCT can be found below. 
```
Finding pattern
Layer 1
	Wire 0, valid, wireN 93, OK in pattern
	Wire 1, valid, wireN 92, OK in pattern, HIT!
	Wire 2, valid, wireN 91, OK in pattern, HIT!
	Wire 3
	Wire 4
Layer 2
	Wire 0
	Wire 1, valid, wireN 92, OK in pattern
	Wire 2, valid, wireN 91, OK in pattern, HIT!
	Wire 3
	Wire 4
Layer 3
	Wire 0
	Wire 1
	Wire 2, valid, wireN 91, OK in pattern, HIT!
	Wire 3
	Wire 4
Layer 4
	Wire 0
	Wire 1
	Wire 2, valid, wireN 91, OK in pattern
	Wire 3, valid, wireN 90, OK in pattern, HIT!
	Wire 4
Layer 5
	Wire 0
	Wire 1
	Wire 2, valid, wireN 91, OK in pattern
	Wire 3, valid, wireN 90, OK in pattern, HIT!
	Wire 4, valid, wireN 89, OK in pattern, HIT!
Layer 6
	Wire 0
	Wire 1
	Wire 2, valid, wireN 91, OK in pattern
	Wire 3, valid, wireN 90, OK in pattern
	Wire 4, valid, wireN 89, OK in pattern, HIT!

Printing hits for keyWire 91 bx 8
Layer 1
	92	91
Layer 2
	91
Layer 3
	91
Layer 4
	90
Layer 5
	90	89
Layer 6
	89
```

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

N/A

@tahuang1991

### CODE REVIEW

**Comment by dildick** [comment]
@rekovic Can you run the tests, please?

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
@rekovic, can you please check and eventually sign?

**Comment by rekovic** [comment]
+1

**Comment by dildick** [comment]
Resolved merged conflict.

**Comment by dildick** [comment]
I'm not sure what happened in the code-checks. I'll check it out.

**Comment by dildick** [comment]
I'll rebase as soon as there is new build.

**Comment by mrodozov** [comment]
code-checks

**Comment by dildick** [comment]
This is after rebasing onto CMSSW_11_2_X_2020-07-08-1100.

**Comment by silviodonato** [comment]
please test
@rekovic

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
ping @rekovic

**Comment by silviodonato** [comment]
merge
@rekovic signed before the conflict https://github.com/cms-sw/cmssw/pull/30301#issuecomment-654853925

**Comment by rekovic** [comment]
+1

File: L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc [hunk 1]
```diff
@@ -295,6 +295,11 @@ void CSCAnodeLCTProcessor::run(const std::vector<int> wire[CSCConstants::NUM_LAY
   // Take the best MAX_CLCTS_PER_PROCESSOR candidates per bx.
   int ALCTIndex_[CSCConstants::MAX_ALCT_TBINS] = {};
 
+  // define a new pattern map
+  // for each key half strip, and for each pattern, store the 2D collection of fired comparator digis
+  std::map<int, std::map<int, CSCCLCTDigi::ComparatorContainer>> hits_in_patterns;
```
Comment (davidlange6) [inline L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc:300]: so this is a map of a map of a vector of vectors?

**Comment by dildick** [comment]
Hold. Not sure how `CSCCLCTDigi::ComparatorContainer` got in CSCALCTProcessor. That needs to be fixed.

File: L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc (hunk 1 above)
Comment (dildick) [reply L1Trigger/CSCTriggerPrimitives/src/CSCAnodeLCTProcessor.cc:300]: @davidlange6 So the type needs to be corrected (https://github.com/cms-sw/cmssw/pull/30940), but it shouldn't have any effect. ComparatorContainer and WireContainer are both 2D vectors.

Yes, it maps wiregroups and pattern numbers onto 2D vectors with hits.

**Pull request merged**